tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala

```diff
@@ -216,27 +261,52 @@ final class ScalaJSClassEmitter(semantics: Semantics, outputMode: OutputMode,
   private def genES6Constructor(tree: LinkedClass): js.Tree = {
     implicit val pos = tree.pos
 
-    val fieldDefs = genFieldDefs(tree)
-    if (fieldDefs.isEmpty && outputMode == OutputMode.ECMAScript6) {
-      js.Skip()
+    if (tree.kind == ClassKind.JSClass) {
+      val js.Function(params, body) = genConstructorFunForJSClass(tree)
+      js.MethodDef(static = false, js.Ident("constructor"), params, body)
     } else {
-      val superCtorCall = tree.superClass.fold[js.Tree] {
-        js.Skip()(tree.pos)
-      } { parentIdent =>
-        js.Apply(js.Super(), Nil)
-      }
-      val initClassData = outputMode match {
-        case OutputMode.ECMAScript6StrongMode =>
-          js.Assign(js.DotSelect(js.This(), js.Ident("$classData")),
-              envField("d", tree.name.name))
-        case _ =>
-          js.Skip()
+      val fieldDefs = genFieldDefs(tree)
+      if (fieldDefs.isEmpty && outputMode == OutputMode.ECMAScript6) {
+        js.Skip()
+      } else {
+        val superCtorCall = tree.superClass.fold[js.Tree] {
+          js.Skip()(tree.pos)
+        } { parentIdent =>
+          js.Apply(js.Super(), Nil)
+        }
+        val initClassData = outputMode match {
+          case OutputMode.ECMAScript6StrongMode =>
+            js.Assign(js.DotSelect(js.This(), js.Ident("$classData")),
+                envField("d", tree.name.name))
+          case _ =>
+            js.Skip()
+        }
+        js.MethodDef(static = false, js.Ident("constructor"), Nil,
+            js.Block(superCtorCall :: initClassData :: fieldDefs))
       }
-      js.MethodDef(static = false, js.Ident("constructor"), Nil,
-          js.Block(superCtorCall :: initClassData :: fieldDefs))
     }
   }
 
+  private def genConstructorFunForJSClass(tree: LinkedClass): js.Function = {
+    implicit val pos = tree.pos
+
+    require(tree.kind == ClassKind.JSClass)
+
+    val ctor = tree.exportedMembers.map(_.tree) collectFirst {
+      case m: MethodDef
+          if m.static == false && m.name == StringLiteral("constructor") =>
+        m
+    } getOrElse {
+      throw new IllegalArgumentException(
+          s"${tree.encodedName} does not have an exported constructor")
+    }
```

Comment:
A class will be exported if it contains a `ExportJSClassDef` in its defs (similar to `ExportModuleDef`). The constructor being exported is normal assuming the class is reachable (by any means). It is completely orthogonal to the class being exported or not.

When exporting a js class, we do not export its constructors. We export the class itself (unlike scala classes)